File: libs/Api/examples/fs20_json.pl
#!/usr/bin/perl -w

# Demo script for switching an actor via JSON API
# requires IPSymcon V3.0+
# see http://www.ip-symcon.de/service/dokumentation/entwicklerbereich/datenaustausch/
# see http://www.tdressler.net/ipsymcon/ipsymcon_api.html
# (C) Thomas Dressler 2013-2014 
# V0.2 04.01.2014

use strict;
use FindBin qw($Bin);
use lib "$Bin:$Bin/../";
use Getopt::Std;
use IPS_JSON;
my %opts;
getopts('H:P:i:s:u:p:c:h',\%opts);
if ($opts{'h'}) {
	usage();
	exit;
}
my $config='IPS_JSON_config.cfg';
$config=$opts{'c'} if $opts{'c'};
our ($host,$port,$user,$password);
#API Access ,see http://www.ip-symcon.de/service/dokumentation/komponenten/verwaltungskonsole/
$host='localhost';
$port=3777;
$user='lizenz@email.ips';
$password='fernzugriff Kennword';

#set config from file if available
if (-r $config) {
	do($config);
}
#commandline parameters have precedence
$host=$opts{'H'} if $opts{'H'};
$port=$opts{'P'} if $opts{'P'};
$user=$opts{'u'} if $opts{'u'};
$password=$opts{'p'} if $opts{'p'};
my $state=$opts{'s'} if exists($opts{s});
my $var_id=$opts{'i'} if $opts{'i'};
if (!($user&& $password)) {
	print "UNKNOWN: need API Username & password\n";
	exit;
}
#reate api url
my $url="http://".$host.":".$port."/api/";#need trailing slash!
my $ips=IPS_JSON->new($url,$user,$password,2);
#if (!$ips->is_string($var_id) || !$ips->is_string($state)) {
if (!defined($var_id) || !defined($state)) {
	print "Error - need atleast FS20 Instance ID and switch state\n";
    usage();
	exit(1);
}

my ($res,$error);
if (!$ips->is_int($var_id)) {
	print "Error VariableID $var_id is not integer\n";
	exit(1);
}

$var_id=int($var_id);#must be integer
$res=$ips->IPS_InstanceExists($var_id);
if (!defined($res)) {
    $error=$ips->getError;
    print "Error - IPS_InstanceExists Request failed:$error\n";
	exit(1);
    
}
if (!$res) {
	print "Error - Instance $var_id doesnt exist on Server $host\n";
	exit(1);
}

my $inst=$ips->IPS_GetInstance($var_id);
if (!$inst) {
    $error=$ips->getError();
    print "Error - IPS_GetInstance Request failed:$error\n";
	exit(1);
}
 
my $name=$inst->{ModuleInfo}{ModuleName};
if ($name eq "FS20") {
    my $action=$ips->bool($state=~/On|Ein|1/i);
    $res=$ips->FS20_SwitchMode($var_id,$action);
    if ($res) {
        print "OK - Set FS20 Switch $var_id to ".(($state=~/On|Ein|1/i)?"On":"Off")." successfully\n";
        exit (0);
    }else{
		$error=$ips->getError();
        print "Error - Set FS20 Switch $var_id to ".(($state=~/On|Ein|1/i)?"On":"Off")." failed:$error\n";
		exit(1);
    }
    
}else{
    print "Error - ID $var_id is not a FS20 instance\n";
	exit(1);
}

sub usage{
    print "usage: $0 -i <instanceid> -s <state> [-c <config_file>]| [-H <ipshost>] [-P <apiport>] [-u <apiuser> -p <apipass>]\n";
    print "state: use On|Ein|1 for 'on', others means 'off'>\n";
}

File: libs/Api/examples/fs20_json.php
<?php
/**
 * @file
 * @brief
 * Demo script for switching an actor via JSON API
 * requires IPSymcon V3.0+
 *
 * @copyright Thomas Dressler 2013-2018
 * @version 0.3
 * @date 25.02.2018
 *
 * @see http://www.ip-symcon.de/service/dokumentation/entwicklerbereich/datenaustausch/
 * @see http://www.tdressler.net/ipsymcon/ipsymcon_api.html

 * @include examples/fs20_json.php
 */
include ('../IPS_JSON.php');
//default configuration
$id=undef;
$host="";
$port="3777";
$user="apiuser";
$password="apipassword";
$config='IPS_JSON_config.cfg';
//read commandline options
$options=getopt("i:H:P:u:p:s:c:h");
//print_r($options);
if (isset($options['h'])) {
	usage();
	exit(0);
}
if (isset($options['c'])) {$config=$options['c'];}
//overwrite default configuration as above with your settings if needed 
if (file_exists($config))  {
	$config_data=file_get_contents($config);
	$res=eval($config_data);
	if ($res===false) {
		print "Error - eval of $config failed. Hint:Use code only, no leading php tags\n";
		exit(1);
	}
}
//commandline parameters have precedence
if (isset($options['P'])) $port=$options['P'];
if (isset($options['H'])) {$host=$options['H'];}
if (isset($options['u'])) {$user=$options['u'];}
if (isset($options['p'])) {$password=$options['p'];}
if (isset($options['i'])) {$id=$options['i'];}
if (isset($options['s'])) {$state=$options['s'];}

if (is_null($id) || is_null($state)) {
	print "Error - need atleast FS20 Instance ID and switch state\n";
    usage();
	exit(0) ;
}

if (!is_numeric($id)){
	print "Error - ID $id looks not as a number\n";
	exit(1);
}
$id=(integer)$id;
if (!class_exists('IPS_JSON'))
{
	print "Error - This requires IPS_JSON class\n";
	exit(1);
}
if (empty($user) || empty($password)) {
	echo "API username or Password missed\n";
	exit;
}
$url="http://".$host.":".$port."/api/";


try {
	/**
	* @var IPS_JSON $ips
	*/
	$ips = new IPS_JSON($url,$user,$password);
	$res=$ips->IPS_InstanceExists($id);
	if (!isset($res)) {
		print "Error - IPS_InstanceExists Request failed:".$ips->getErrorMessage()."\n";
		exit(1);
	}
	if ($res===false) {
		print "Error - Instance $id doesnt exists\n";
		exit(1);
	}
	$inst=$ips->IPS_GetInstance($id);
	if (!$inst) {
		print "Error - IPS_GetInstance Request failed=".$ips->getErrorMessage()."\n";
		exit(1);
	}
	$name=$inst['ModuleInfo']['ModuleName'];
	if ($name == "FS20") {
		$switch=preg_match("/On|Ein|1/i",$state)?true:false;
		$res=$ips->FS20_SwitchMode($id,$switch);
		if (!$res) {
			print "Error - Set FS20 Switch $id to ".($switch?"On":"Off")." failed=".$ips->getErrorMessage()."\n";
			exit(1);
			
		}else{
			print "OK - Set FS20 Switch $id to ".($switch?"On":"Off")." successfully\n";
		}
	}else{
		print "Error - Instance $id is not an instance of FS20 module\n";
		exit(1);
	}

}catch (Exception $e) {
	echo "Error - Request failed->";
	echo $e->getMessage()."\n";
	exit(1);
}
exit (0);


/**
 * print usage information
 */
function usage(){
    print "usage: ".basename(__FILE__)." -i <instanceid> -s <state> [-h] [-c <config_file>] | [-H <ipshost> -P <apiport> -u <apiuser> -p <apipass>] \n";
    print "state: use On|Ein|1 for 'on', others means 'off'>\n";
}
?>